src/ray/core_worker/transport/concurrency_group_manager.cc

```diff
@@ -81,9 +88,54 @@ std::shared_ptr<ExecutorType> ConcurrencyGroupManager<ExecutorType>::GetDefaultE
   return default_executor_;
 }
 
+template <typename ExecutorType>
+std::optional<std::function<void()>>
+ConcurrencyGroupManager<ExecutorType>::InitializeExecutor(
+    std::shared_ptr<ExecutorType> executor) {
+  if (language_ == Language::PYTHON) {
+    if constexpr (std::is_same<ExecutorType, BoundedExecutor>::value) {
+      PyGILState_STATE gstate;
+      PyThreadState *tstate;
+      // Create a promise/future pair to synchronize the initialization
+      std::promise<void> init_promise;
+      auto init_future = init_promise.get_future();
+
+      executor->Post([&gstate, &tstate, &init_promise]() {
+        // `PyGILState_Ensure()` makes this C++ thread appear as a Python thread
+        // from the perspective of the Python interpreter, regardless of whether
+        // the thread is executing Python code (i.e., Ray tasks or actors) or not.
+        gstate = PyGILState_Ensure();
+        // Release the GIL so the main thread can acquire it to execute the
+        // control plane logic.
+        tstate = PyEval_SaveThread();
+        init_promise.set_value();
+      });
+
+      // Wait for Python initialization to complete
+      init_future.wait();
+
+      return [&gstate, &tstate, executor]() {
+        executor->Post([&gstate, &tstate]() {
+          PyEval_RestoreThread(tstate);
+          PyGILState_Release(gstate);
+        });
+      };
+    }
+    // TODO: Consider whether we need to handle FiberState or using BoundedExecutor for
+    // both sync and async cases.
+  }
+  return std::nullopt;
+}
```

Comment:
I don't think we should be hard-coding anything about language-level initialization here.

Is it not possible to pass an `initialize_thread` callback from Python into cpp?

Reply:
It should be possible, but what's the motivation if the initialization function can be determined by the language? In addition, the function call path will become longer, making it more prone to issues.

One advantage of passing the callback, as I see it, is that it avoids using `#ifdef LANGUAGE ...` to exclude Python.h when the frontend is not Python.

Reply:
> the function call path will become longer, making it more prone to issues

can you explain what this means?

The main motivation is from a higher-level design standpoint, we should not be injecting dependencies or assumptions about the language frontend into the core of the system. It's an abstraction leak and is going to make maintaining and modifying the code harder going forward.